# Introduction to data analysis in Python

Based on Software Carpentry's "Programming with Python" https://software-carpentry.org/lessons/ and Data Carpentry's "Data Analysis and Visualization in Python for Ecologists" https://datacarpentry.org/lessons/

Recommended setup: Anaconda / miniconda on Linux or Mac (Windows Subsystem for Linux if on Windows).


## Datatypes, variables methods:

### Questions
* How do I program in Python?

* How can I represent my data in Python?

### Objectives

* Define the following data types in Python: strings, integers, and floats.

* Perform mathematical operations in Python using basic operators.

* Define the following as it relates to Python: lists, tuples, and dictionaries.

### Variables

Any Python interpreter can be used as a calculator:




In [1]:
2 + 2

4

In [3]:
6 * 8

48

This is great but not very interesting. To do anything useful with data, we need to assign its value to a variable. In Python, we can assign a value to a variable, using the equals sign `=`. For example, we can track the weight of a patient who weighs 60 kilograms by assigning the value 60 to a variable `weight_kg`:

In [2]:
weight_kg = 60

From now on, whenever we use `weight_kg`, Python will substitute the value we assigned to it. In layman’s terms, a variable is a name for a value.

In Python, variable names:

* can include letters, digits, and underscores
* cannot start with a digit
* are case sensitive.

This means that, for example:

* `weight0` is a valid variable name, whereas `0weight` is not
* `weight` and `Weight` are different variables

### Built-in Python functions

To carry out common tasks with data and variables in Python, the language provides us with several built-in functions. To display information to the screen, we use the print function:

In [ ]:
print(weight_lb)

132.66
inflam_001


When we want to make use of a function, referred to as calling the function, we follow its name by parentheses. The parentheses are important: if you leave them off, the function doesn’t actually run! Sometimes you will include values or variables inside the parentheses for the function to use. In the case of print, we use the parentheses to tell the function what value we want to display. We will learn more about how functions work and how to create our own in later episodes.

We can display multiple things at once using only one `print` call:

In [9]:
print('Weight in kilograms:', weight_kg)

Weight in kilograms: 60.3


Moreover, we can do arithmetic with variables right inside the print function:

In [ ]:
print('weight in pounds:', 2.2 * weight_kg)

weight in pounds: 132.66


The above command, however, did not change the value of weight_kg:

In [ ]:
print(weight_kg)

60.3


To change the value of the `weight_kg` variable, we have to assign `weight_kg` a new value using the equals = sign:

In [ ]:
weight_kg = 65.0
print('weight in kilograms is now:', weight_kg)

weight in kilograms is now: 65.0


### Built-in data types

**Strings, integers, and floats**

Python knows various types of data. Three common ones are:

* integer numbers
* floating point numbers, and
* strings.

In the example above, variable `weight_kg` has an integer value of 60. If we want to more precisely track the weight of our patient, we can use a floating point value by executing:

In [5]:
weight_kg = 60.3

To create a string, we add single or double quotes around some text. To identify and track a patient throughout our study, we can assign each person a unique identifier by storing it in a string:

In [6]:
patient_id = '001'

We can also call a function to check the type of a variable:

In [10]:
type(weight_kg)

float

In [11]:
type(patient_id)

str

**Sequences: Lists and Tuples**

Lists are a common data structure to hold an ordered sequence of elements. Each element can be accessed by an index. Note that Python indexes start with 0 instead of 1:


In [7]:
numbers = [1, 2, 3]
numbers[0]

1

A `for` loop can be used to access the elements in a list or other Python data structure one at a time:

In [8]:
for num in numbers:
    print(num)

1
2
3


**Indentation** is very important in Python. Note that the second line in the example above is indented.

To add elements to the end of a list, we can use the `append` method. Methods are a way to interact with an object (a list, for example). We can invoke a method using the dot . followed by the method name and a list of arguments in parentheses. Let’s look at an example using `append`:

In [12]:
numbers.append(4)
print(numbers)

[1, 2, 3, 4]


To find out what methods are available for an object, we can use the built-in `help` command:

In [13]:
help(numbers)

Help on list object:

class list(object)
 |  list(iterable=(), /)
 |  
 |  Built-in mutable sequence.
 |  
 |  If no argument is given, the constructor creates a new empty list.
 |  The argument must be an iterable if specified.
 |  
 |  Methods defined here:
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __delitem__(self, key, /)
 |      Delete self[key].
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getitem__(...)
 |      x.__getitem__(y) <==> x[y]
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __iadd__(self, value, /)
 |      Implement self+=value.
 |  
 |  __imul__(self, value, /)
 |      Implement self*=value.
 |  
 |  __init__(self, /, *args, **kwargs)
 |      Initialize self.  See help(type(self)) for accurate sign

A **tuple** is similar to a list in that it’s an ordered sequence of elements. However, tuples can not be changed once created (they are “immutable”). Tuples are created by placing comma-separated values inside parentheses ().

In [14]:
# Tuples use parentheses
a_tuple = (1, 2, 3)
another_tuple = ('blue', 'green', 'red')

# Note: lists use square brackets
a_list = [1, 2, 3]

A **dictionary** is a container that holds pairs of objects - keys and values.

In [15]:
translation = {'one': 'first', 'two': 'second'}
translation['one']

'first'

Dictionaries work a lot like lists - except that you index them with *keys*. You can think about a key as a name or unique identifier for the value it corresponds to.

In [16]:
rev = {'first': 'one', 'second': 'two'}
rev['first']

'one'

To add an item to the dictionary we assign a value to a new key:

In [17]:
rev = {'first': 'one', 'second': 'two'}
rev['third'] = 'three'
rev

{'first': 'one', 'second': 'two', 'third': 'three'}

Using `for` loops with dictionaries is a little more complicated. We can do this in two ways:

In [18]:
for key, value in rev.items():
    print(key, '->', value)

first -> one
second -> two
third -> three


In [19]:
for key in rev.keys():
    print(key, '->', rev[key])

first -> one
second -> two
third -> three


### Functions

Defining a section of code as a **function** in Python is done using the `def` keyword. For example a function that takes two arguments and returns their sum can be defined as:

In [1]:
def add_function(a, b):
    result = a + b
    return result

z = add_function(20, 22)
print(z)

42


### Conditionals

We can ask Python to take different actions, depending on a condition, with an `if` statement:

In [2]:
num = 37
if num > 100:
    print('greater')
else:
    print('not greater')
print('done')

not greater
done


The second line of this code uses the keyword `if` to tell Python that we want to make a choice. If the test that follows the `if` statement is true, the body of the `if` (i.e., the set of lines indented underneath it) is executed, and “greater” is printed. If the test is false, the body of the else is executed instead, and “not greater” is printed. Only one or the other is ever executed before continuing on with program execution to print “done”.

Conditional statements don’t have to include an `else`. If there isn’t one, Python simply does nothing if the test is false:

In [3]:
num = 53
print('before conditional...')
if num > 100:
    print(num, 'is greater than 100')
print('...after conditional')

before conditional...
...after conditional


We can also chain several tests together using `elif`, which is short for “else if”. The following Python code uses `elif` to print the sign of a number.

In [4]:
num = -3

if num > 0:
    print(num, 'is positive')
elif num == 0:
    print(num, 'is zero')
else:
    print(num, 'is negative')

-3 is negative


### Boolean statements

Along with the > and == operators we have already used for comparing values in our conditionals, there are a few more options to know about:

- `>`: greater than
- `<`: less than
- `==`: equal to
- `!=`: does not equal
- `>=`: greater than or equal to
- `<=`: less than or equal to

We can also combine tests using `and` and `or`. `and` is only true if both parts are true:

In [5]:
if (1 > 0) and (-1 >= 0):
    print('both parts are true')
else:
    print('at least one part is false')

at least one part is false


while `or` is true if at least one part is true:

In [6]:
if (1 < 0) or (1 >= 0):
    print('at least one test is true')

at least one test is true


Sometimes it is useful to check whether some condition is not true. The Boolean operator `not` can do this explicitly

In [11]:
if not (1 < 0):
    print('1 is not smaller than 0')

1 is not smaller than 0


### Importing libraries

Importing a library is like getting a piece of lab equipment out of a storage locker and setting it up on the bench. Libraries provide additional functionality to the basic Python package, much like a new piece of equipment adds functionality to a lab space. Just like in the lab, importing too many libraries can sometimes complicate and slow down your programs - so we only import what we need for each program.

Once we’ve imported the library, we can ask the library to perform functions that are not built-in:

In [ ]:
import numpy
weight_kg = 60.6
print(numpy.round(weight_kg))


61.0


It is common to rename libraries to appreviated names:

In [ ]:
import numpy as np
print(np.round(weight_kg))

61.0


### Exercises

What values do the variables mass and age have after each of the following statements? Test your answer by executing the lines.

In [ ]:
mass = 47.5
age = 122
mass = mass * 2.0
age = age - 20

Python allows you to assign multiple values to multiple variables in one line by separating the variables and values with commas. What does the following program print out?

In [ ]:
first, second = 'Grace', 'Hopper'
third, fourth = second, first
print(third, fourth)

What are the data types of the following variables?

In [ ]:
planet = 'Earth'
apples = 5
distance = 10.5

In [8]:
a_list = [1, 2, 3]
a_tuple = (1, 2, 3)

Change the value of the second element in both the list and tuple to 5:

Write a function `square` that takes a `list` or `tuple` as its argument and returns a respective `list` or `tuple` containing the squares of the values in the original list.

Write a function `is_number` that takes an argument and returns whether the argument is a number.

### Key points

* Basic data types in Python include integers, strings, and floating-point numbers.

* Use `variable = value` to assign a value to a variable in order to record it in memory.

* Variables are created on demand whenever a value is assigned to them.

* Built-in functions are always available to use.

* Lists and tuples are similar in that they are ordered lists of elements; they differ in that a tuple is immutable (cannot be changed).

* Dictionaries are data structures that provide mappings between keys and values.

* Use for variable in sequence to process the elements of a sequence one at a time.

* Use if condition to start a conditional statement, elif condition to provide additional tests, and else to provide a default.



## Data analysis using Pandas

### Questions

- How can I import data in Python?
- What is Pandas?
- Why should I use Pandas to work with data?

### Objectives

- Navigate the workshop directory and download a dataset.
- Explain what a library is and what libraries are used for.
- Describe what the Python Data Analysis Library (Pandas) is.
- Load the Python Data Analysis Library (Pandas).
- Use read_csv to read tabular data into Python.
- Describe what a DataFrame is in Python.
- Access and summarize data stored in a DataFrame.
- Define indexing as it relates to data structures.
- Perform basic mathematical operations and summary statistics on data in a Pandas DataFrame.
- Create simple plots.

### Working With Pandas DataFrames in Python

We can automate the process of performing data manipulations in Python. It’s efficient to spend time building the code to perform these tasks because once it’s built, we can use it over and over on different datasets that use a similar format. This makes our methods easily reproducible. We can also easily share our code with colleagues and they can replicate the same analysis.

### Our Data

For this lesson, we will be real data from traffic registration points in Norway.

We are studying (...). The dataset is stored as a .csv file: each row holds information for a single traffic event.

The first few rows of our first file look like this:

```
_id;_index;_score;_type;county_id;created_at_timestamp;datalogger_type;event_emitted_timestamp;event_number;event_timestamp;firmware_version;klokketime;lane;length;qspeed;region_id;speed;time_gap;traffic_registration_point_id;valid_classification;valid_event;valid_length;valid_speed;vehicle_type;vehicle_type_quality;vehicle_type_raw;weight;with_traffic_registration_point_direction;wrong_direction							
1-27849732;traffic_event_vehicle_2020_10;;traffic_event_vehicle;18;2020-10-26	 02:16:08.090;EMU;2020-10-25	 23:59:53.564;27849732;2020-10-25	 23:59:53.564;1.04 EMU3/15606;23;4;4	5;0;1;73	2;123	9;1;true;true;true;true;2;;LMV2;0;false;false	
1-27849731;traffic_event_vehicle_2020_10;;traffic_event_vehicle;18;2020-10-26	 02:16:08.116;EMU;2020-10-25	 23:58:28.296;27849731;2020-10-25	 23:58:28.296;1.04 EMU3/15606;23;3;4	6;0;1;75	4;1	8;1;true;true;true;true;2;;LMV2;0;true;false	
1-27849730;traffic_event_vehicle_2020_10;;traffic_event_vehicle;18;2020-10-26	 02:16:08.066;EMU;2020-10-25	 23:58:25.527;27849730;2020-10-25	 23:58:25.527;1.04 EMU3/15606;23;3;4	6;0;1;80	1;6	8;1;true;true;true;true;2;;LMV2;0;true;false	
1-27849729;traffic_event_vehicle_2020_10;;traffic_event_vehicle;18;2020-10-26	 02:16:08.085;EMU;2020-10-25	 23:58:19.077;27849729;2020-10-25	 23:58:19.077;1.04 EMU3/15606;23;3;4	1;0;1;75	4;69	9;1;true;true;true;true;2;;LMV2;0;true;false	
```



### Pandas in Python

One of the best options for working with tabular data in Python is to use the Python Data Analysis Library (a.k.a. Pandas). The Pandas library provides data structures, produces high quality plots with matplotlib and integrates nicely with other libraries that use NumPy (which is another Python library) arrays.

Python doesn’t load all of the libraries available to it by default. We have to add an import statement to our code in order to use library functions:

In [2]:
import pandas as pd

### Reading CSV Data Using Pandas

We will begin by locating and reading our data which are in CSV format. CSV stands for Comma-Separated Values and is a common way to store formatted data. Other symbols may also be used, so you might see tab-separated, colon-separated or space separated files. It is quite easy to replace one separator with another, to match your application. The first line in the file often has headers to explain what is in each column. CSV (and other separators) make it easy to share data, and can be imported and exported from many applications, including Microsoft Excel. For more details on CSV files, see the Data Organisation in Spreadsheets lesson. We can use Pandas’ `read_csv` function to pull the file directly into a `DataFrame`.

### So What’s a DataFrame?

A DataFrame is a 2-dimensional data structure that can store data of different types (including characters, integers, floating point values, factors and more) in columns. It is similar to a spreadsheet or an SQL table or the `data.frame` in R. A DataFrame always has an index (0-based). An index refers to the position of an element in the data structure.

In [21]:
pd.read_csv('/content/trafikkregistreringspunkt_1_-_uke_42_og_43_2020.csv')

created_at_timestamp  ... wrong_direction
0       2020-10-26, 02:16:08.090  ...           False
1       2020-10-26, 02:16:08.116  ...           False
2       2020-10-26, 02:16:08.066  ...           False
3       2020-10-26, 02:16:08.085  ...           False
4       2020-10-26, 02:16:08.116  ...           False
...                          ...  ...             ...
299772  2020-10-12, 03:06:45.940  ...           False
299773  2020-10-12, 03:06:45.939  ...           False
299774  2020-10-12, 03:06:45.944  ...           False
299775  2020-10-12, 03:06:45.947  ...           False
299776  2020-10-12, 03:06:45.940  ...           False

[299777 rows x 22 columns]

We can see that there were 299,777 rows parsed. Each row has 29 columns. The first column is the index of the DataFrame. The index is used to identify the position of the data, but it is not an actual column of the DataFrame. It looks like the `read_csv` function in Pandas read our file properly. However, we haven’t saved any data to memory so we can work with it. We need to assign the DataFrame to a variable. Remember that a variable is a name for a value, such as `x`, or `data`. We can create a new object with a variable name by assigning a value to it using `=`.

Let’s call the imported survey data `data`:

In [3]:
data = pd.read_csv('/content/trafikkregistreringspunkt_1_-_uke_42_og_43_2020.csv')

Notice when you assign the imported DataFrame to a variable, Python does not produce any output on the screen. We can view the value of the surveys_df object by typing its name into the Python command prompt.

In [15]:
data

_id  ... wrong_direction
0       1-27849732  ...           False
1       1-27849731  ...           False
2       1-27849730  ...           False
3       1-27849729  ...           False
4       1-27849728  ...           False
...            ...  ...             ...
299772  1-27549743  ...           False
299773  1-27549742  ...           False
299774  1-27549741  ...           False
299775  1-27549740  ...           False
299776  1-27549739  ...           False

[299777 rows x 29 columns]

which prints contents like above.

### Exploring the data

Again, we can use the `type` function to see what kind of thing `data` is:

In [22]:
type(data)

pandas.core.frame.DataFrame

As expected, it’s a DataFrame (or, to use the full name that Python uses to refer to it internally, a pandas.core.frame.DataFrame).

What kind of data does `data` contain? DataFrames have an attribute called `dtypes` that answers this:

In [23]:
data.dtypes

created_at_timestamp                          object
event_emitted_timestamp                       object
event_number                                   int64
event_timestamp                               object
klokketime                                     int64
lane                                           int64
length                                       float64
qspeed                                       float64
region_id                                      int64
speed                                        float64
time_gap                                     float64
traffic_registration_point_id                  int64
valid_classification                            bool
valid_event                                     bool
valid_length                                    bool
valid_speed                                     bool
vehicle_type                                 float64
vehicle_type_quality                         float64
vehicle_type_raw                              

Not all the values in a column have the same type. For example, `klokketime` have type `int64`, which is a kind of integer. Cells in the `klokketime` column cannot have fractional values, but the `_score` and `vehicle_type` columns can, because they have type `float64`. The `object` type doesn’t have a very helpful name, but it can for example represent `Strings` like `created_at_timestamp`.

### Useful Ways to View DataFrame objects in Python

There are many ways to summarize and access the data stored in DataFrames, using attributes and methods provided by the DataFrame object.

To access an attribute, use the DataFrame object name followed by the attribute name `df_object.attribute`. Using the DataFrame `data` and attribute `columns`, an index of all the column names in the DataFrame can be accessed with `data.columns`.

Methods are called in a similar fashion using the syntax `df_object.method()`. As an example, `data.head()` gets the first few rows in the DataFrame `data` using the `head()` method. With a method, we can supply extra information in the parens to control behaviour.

Using our DataFrame `data`, try out the attributes & methods below to see what they return.
- `data.columns`
- `data.shape` Take note of the output of shape - what format does it return the shape of the DataFrame in?
- `data.head()` Also, what does `data.head(15)` do?
- `data.tail()`

### Calculating Statistics From Data In A Pandas DataFrame

We’ve read our data into Python. Next, let’s perform some quick summary statistics to learn more about the data that we’re working with. We might want to know how many animals were collected in each site, or how many of each species were caught. We can perform summary stats quickly using groups. But first we need to figure out what we want to group by.

Let’s begin by exploring our data:

In [24]:
data.columns

Index(['created_at_timestamp', 'event_emitted_timestamp', 'event_number',
       'event_timestamp', 'klokketime', 'lane', 'length', 'qspeed',
       'region_id', 'speed', 'time_gap', 'traffic_registration_point_id',
       'valid_classification', 'valid_event', 'valid_length', 'valid_speed',
       'vehicle_type', 'vehicle_type_quality', 'vehicle_type_raw', 'weight',
       'with_traffic_registration_point_direction', 'wrong_direction'],
      dtype='object')

Let’s get a list of all the vehicle types. The `unique` function tells us all of the unique values in the `vehicle_type_raw` column.

In [25]:
data['vehicle_type_raw'].unique()

array(['LMV2', 'HMV', 'LMV2+WC', 'LMV1', 'UC LOOP'], dtype=object)

**Exercise**

What does `data['vehicle_type_raw'].nunique()` give?

Create a list of unique vehicle types (`vehicle_type_raw`) found in the data. Call it `vehicle_types`.

5

We can easily calculate basic statistics for all numerical records:

In [6]:
data.describe()

event_number     klokketime  ...  vehicle_type_quality    weight
count  2.997770e+05  299777.000000  ...                   0.0  299777.0
mean   2.769971e+07      13.560760  ...                   NaN       0.0
std    8.662735e+04       4.606718  ...                   NaN       0.0
min    2.754974e+07       0.000000  ...                   NaN       0.0
25%    2.762468e+07      10.000000  ...                   NaN       0.0
50%    2.769963e+07      14.000000  ...                   NaN       0.0
75%    2.777479e+07      17.000000  ...                   NaN       0.0
max    2.784973e+07      23.000000  ...                   NaN       0.0

[8 rows x 12 columns]

### Grouping data

We often want to calculate summary statistics grouped by subsets or attributes within fields of our data. For example, we might want to calculate the average speed of all vehicles per vehicle type.

Similarly to calculating statistics for the whole dataset we can calculate them for a single column:



In [7]:
data['speed'].describe()

count    299777.000000
mean         76.711151
std          14.752190
min           2.000000
25%          71.200000
50%          77.700000
75%          82.700000
max        6553.500000
Name: speed, dtype: float64

We can also extract one specific metric if we wish:

In [8]:
data['speed'].min()
data['speed'].max()
data['speed'].mean()
data['speed'].std()
data['speed'].count()

299777

But if we want to summarize by one or more variables, for example vehicle type, we can use Pandas’ `.groupby` method. Once we’ve created a groupby DataFrame, we can quickly calculate summary statistics by a group of our choice.

In [9]:
grouped_data = data.groupby('vehicle_type')

In [10]:
grouped_data.describe()

event_number                              ... weight          
                    count          mean           std  ...    50%  75%  max
vehicle_type                                           ...                 
1.0                 349.0  2.764940e+07  80159.023739  ...    0.0  0.0  0.0
2.0              276771.0  2.770051e+07  86710.704202  ...    0.0  0.0  0.0
3.0               16683.0  2.768975e+07  84655.463763  ...    0.0  0.0  0.0
4.0                5940.0  2.769364e+07  85767.154850  ...    0.0  0.0  0.0

[4 rows x 88 columns]

In [11]:
grouped_data.mean()

event_number  ...  wrong_direction
vehicle_type                ...                 
1.0           2.764940e+07  ...         0.000000
2.0           2.770051e+07  ...         0.000011
3.0           2.768975e+07  ...         0.000120
4.0           2.769364e+07  ...         0.000000

[4 rows x 17 columns]

**Exercise**

1. How many entries do we have per vehicle type?

2. What happens when you group by two columns using the following syntax and then calculate mean values?

```
grouped_data2 = data.groupby(['vehicle_type', 'lane'])
```

